pkg/proxy/iptables/proxier.go

```diff
@@ -0,0 +1,578 @@
+				state = newServiceInfo(svcPort)
+				proxier.serviceMap[svcPort] = state
+			}
+			curEndpoints := []string{}
+			if state != nil {
+				curEndpoints = state.endpoints
+			}
+			newEndpoints := flattenValidEndpoints(portsToEndpoints[portname])
+
+			if len(curEndpoints) != len(newEndpoints) || !slicesEquiv(slice.CopyStrings(curEndpoints), newEndpoints) {
+				glog.V(1).Infof("Setting endpoints for %s to %+v", svcPort, newEndpoints)
+				state.endpoints = newEndpoints
+			}
+			registeredEndpoints[svcPort] = true
+		}
+	}
+	// Remove endpoints missing from the update.
+	for service := range proxier.serviceMap {
+		if _, exists := registeredEndpoints[service]; !exists {
+			glog.V(2).Infof("Removing endpoints for %s", service)
+			// Set the endpoints to nil, we will check for this in OnServiceUpdate so that we
+			// only remove ServicePorts that have no endpoints and were not in the service update,
+			// that way we only remove ServicePorts that were not in both.
+			proxier.serviceMap[service].endpoints = nil
+		}
+	}
+}
+
+// used in OnEndpointsUpdate
+type hostPortPair struct {
+	host string
+	port int
+}
+
+func isValidEndpoint(hpp *hostPortPair) bool {
+	return hpp.host != "" && hpp.port > 0
+}
+
+// Tests whether two slices are equivalent.  This sorts both slices in-place.
+func slicesEquiv(lhs, rhs []string) bool {
+	if len(lhs) != len(rhs) {
+		return false
+	}
+	if reflect.DeepEqual(slice.SortStrings(lhs), slice.SortStrings(rhs)) {
+		return true
+	}
+	return false
+}
+
+func flattenValidEndpoints(endpoints []hostPortPair) []string {
+	// Convert Endpoint objects into strings for easier use later.
+	var result []string
+	for i := range endpoints {
+		hpp := &endpoints[i]
+		if isValidEndpoint(hpp) {
+			result = append(result, net.JoinHostPort(hpp.host, strconv.Itoa(hpp.port)))
+		}
+	}
+	return result
+}
+
+// servicePortToServiceChain takes the ServicePortName for a
+// service and returns the associated iptables chain
+// this is computed by hashing (sha256) then encoding to base64 and
+// truncating with the prefix "KUBE-SVC-"
+// We do this because Iptables Chain Names must be <= 28 chars long
+func servicePortToServiceChain(s proxy.ServicePortName) utiliptables.Chain {
+	hash := sha256.Sum256([]byte(s.String()))
+	encoded := base32.StdEncoding.EncodeToString(hash[:])
+	return utiliptables.Chain("KUBE-SVC-" + encoded[:19])
+}
+
+// this is the same as servicePortToServiceChain but with the endpoint included essentially
+func servicePortAndEndpointToServiceChain(s proxy.ServicePortName, endpoint string) utiliptables.Chain {
+	hash := sha256.Sum256([]byte(s.String() + "_" + endpoint))
+	encoded := base32.StdEncoding.EncodeToString(hash[:])
+	return utiliptables.Chain("KUBE-SVC-" + encoded[:19])
+}
+
+// This is where all of the iptables-save/restore calls happen.
+// The only other iptables rules are those that are setup in iptablesInit()
+// assumes proxier.mu is held
+func (proxier *Proxier) syncProxyRules() error {
```

Comment:
Ok, sounds like a good quick fix.

On Wed, Aug 12, 2015 at 1:35 AM, Tim Hockin <notifications@github.com>
wrote:

> In pkg/proxy/iptables/proxier.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/9210#discussion_r36827879>
> :
>
> > +	hash := sha256.Sum256([]byte(s.String()))
> > +	encoded := base32.StdEncoding.EncodeToString(hash[:])
> > +	return utiliptables.Chain("KUBE-SVC-" + encoded[:19])
> > +}
> > +
> > +// this is the same as servicePortToServiceChain but with the endpoint included essentially
> > +func servicePortAndEndpointToServiceChain(s proxy.ServicePortName, endpoint string) utiliptables.Chain {
> > +	hash := sha256.Sum256([]byte(s.String() + "_" + endpoint))
> > +	encoded := base32.StdEncoding.EncodeToString(hash[:])
> > +	return utiliptables.Chain("KUBE-SVC-" + encoded[:19])
> > +}
> > +
> > +// This is where all of the iptables-save/restore calls happen.
> > +// The only other iptables rules are those that are setup in iptablesInit()
> > +// assumes proxier.mu is held
> > +func (proxier *Proxier) syncProxyRules() error {
>
> Seeing interesting behavior: at startup it detects all chains as
> to-be-removed because it got an update for Services but not endpoint. Then
> it gets an endpoints update and restores them all.
>
> I think syncRules should check that it has received both types of updates
> before it tries to sync. Just set two flags on proxier and set them to true
> in the respective On*Update() callbacks.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/9210/files#r36827879>
> .
>